modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageCompactionTriggerConfiguration.java

```diff
@@ -0,0 +1,130 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.metastorage.impl;
+
+import static org.apache.ignite.internal.util.CompletableFutures.nullCompletedFuture;
+
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.ignite.internal.configuration.SystemDistributedConfiguration;
+import org.apache.ignite.internal.configuration.SystemDistributedView;
+import org.apache.ignite.internal.configuration.SystemPropertyView;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.jetbrains.annotations.Nullable;
+
+/** Configuration for metastorage compaction triggering based on distributed system properties. */
+public class MetaStorageCompactionTriggerConfiguration {
+    private static final IgniteLogger LOG = Loggers.forClass(MetaStorageCompactionTriggerConfiguration.class);
+
+    /**
+     * System property that defines compaction start interval (in milliseconds).
+     *
+     * <p>Default value is {@link #INTERVAL_DEFAULT_VALUE}.</p>
+     */
+    public static final String INTERVAL_SYSTEM_PROPERTY_NAME = "metastorageCompactionInterval";
+
+    /**
+     * System property that defines compaction data availability time (in milliseconds).
+     *
+     * <p>Default value is {@link #DATA_AVAILABILITY_TIME_DEFAULT_VALUE}.</p>
+     */
+    public static final String DATA_AVAILABILITY_TIME_SYSTEM_PROPERTY_NAME = "metastorageCompactionDataAvailabilityTime";
+
+    /** Default value for the {@link #INTERVAL_SYSTEM_PROPERTY_NAME}. */
+    // TODO: IGNITE-23280 Make default 1 minute
+    public static final long INTERVAL_DEFAULT_VALUE = Long.MAX_VALUE;
+
+    /** Default value for the {@link #DATA_AVAILABILITY_TIME_SYSTEM_PROPERTY_NAME}. */
+    // TODO: IGNITE-23280 Make default 1 hour
+    public static final long DATA_AVAILABILITY_TIME_DEFAULT_VALUE = Long.MAX_VALUE;
+
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+
+    /** Guarded by {@link #rwLock}. */
+    private long interval;
+
+    /** Guarded by {@link #rwLock}. */
+    private long dataAvailabilityTime;
+
+    /** Constructor. */
+    public MetaStorageCompactionTriggerConfiguration(SystemDistributedConfiguration systemDistributedConfiguration) {
+        updateSystemProperties(systemDistributedConfiguration.value());
+
+        systemDistributedConfiguration.listen(ctx -> {
```

Comment:
Is a race between the initial update and subscribing to events possible?

Reply:
I don't think so, we will use it in the constructor when creating a trigger, and we will receive a notification after deploying watches.

Reply:
Please add a comment explaining this